### ISSUE TITLE
Clarify usage of 412 and 410 in the participant enlistment cases

### ISSUE THREAD

**Issue by xstefank** [eclipse/microprofile-lra#290]
We have to distinguish two cases:

* participant returning values from Compensate, Complete, Status
  * 410 means participant has no knowledge of received LRA (basically same as Compensated/Completed) -- [spec#302]( https://github.com/eclipse/microprofile-lra/blame/master/spec/src/main/asciidoc/microprofile-lra-spec.adoc#L302)
  * 412 is not allowed

* values returned from user invocations of LRA methods with included LRA context 
  * this needs clarification as per https://github.com/eclipse/microprofile-lra/pull/281

The options discussed for "values returned from user invocations of LRA methods with included LRA context":

1. use both 412 and 410
* 412 LRA context is present when it shouldn't or not present when it should or invalid context (see MANDATORY or NEVER types) and invalid received URI
* 410 expired context (valid URI but not found in the implementation resources)
  * basically closed or canceled LRA already removed from the implementation resources

2. use only 412 as it is a different use case then participant returning values
* 412 means that the context is present when it shouldn't or not present when it should **or** that the received context is expired (so no distinction) or invalid URI

Can we agree on one of the options and unify the text and test expectations across the specification and the Javadoc?

**Comment by xstefank** [comment]
@rdebusscher @mmusgrov @ochaloup can we make some progress on this issue please? This is one of the last outstanding things for 1.0 release.

**Comment by mmusgrov** [comment]
410 Gone means the implementation either does not know about this LRA or knows it has already finished.

412 Precondition Failed means that the method invocation failed because the implementation could not fulfil one of the conditions required by the spec (usually due to application error). We explicitly specify the conditions under which 412 should be returned.

**Comment by rdebusscher** [comment]
Agree with 2 (which is online with @mmusgrov comment)

But since you never can have the situation that implementation doesn't know about an LRA during enlistment, 410 makes only sense for callback and status calls.

For those callback and status call, 412 is probably not required (but should have a detailed look at all possible scenarios)

**Comment by xstefank** [comment]
TL;DR;

If I've understood correctly then @mmusgrov is opting for option 1 and @rdebusscher is opting for option 2. Can we reach some form of compromise? I personally don't have a strong opinion here but if we are to vote on this issue than I can vote.

--------------------

**My interpretation**: We are talking only about user invoking LRA method and the status codes that can be returned to the invoking user.

412 is defined to be returned in situations that are described in [the Javadoc of LRA annotation](https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.java). E.g., [@LRA(MANDATORY) endpoint called without LRA header](https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/ws/rs/LRA.java#L206) so let's take it as an example.

IIUIC, 410 returned from such endpoint would mean that the LRA header **IS** present but the LRA is not tracked (active, ended, or something else). So let's describe individual cases:

Option 1:
* call `@LRA(MANDATORY)` **with** LRA header == _200_
* call `@LRA(MANDATORY)` **without** LRA header == _412_
* call `@LRA(MANDATORY)` **with** LRA header which has been closed somewhere else (I as a caller don't know about it) == _410_ # which is a clear indication of the use case but of course **now the caller needs to check on every invocation whether 412 or 410 was returned**
* call `@LRA(MANDATORY)` **with** LRA header **containing wrong content (e.g. asdfasdf)** == 412

Option 2:
* call `@LRA(MANDATORY)` **with** LRA header == _200_
* call `@LRA(MANDATORY)` **without** LRA header == _412_
* call `@LRA(MANDATORY)` **with** LRA header which has been closed somewhere else (I as a caller don't know about it) == _412_ # this may be confusing as I am acting according to the specification (including header which I think is active) but the specification returning 412 is saying that header is not present. Of course, we can change the text to accommodate this case so **the user knows that the LRA header was present and thus receiving 412 means that it must have expired as I know that the header was included**
* call `@LRA(MANDATORY)` **with** LRA header **containing wrong content (e.g. asdfasdf)** == 412


So I am still undecided whether it is important for users to know that the passed header is wrong for the price of having two failure checks with each REST LRA method invocations.

**Comment by rdebusscher** [comment]
>call @LRA(MANDATORY) with LRA header which has been closed somewhere

Spec says that it needs to be a valid, active LRA. So 412 is not 'wrong' as the precondition is violated (an active valid LRA). 412 does not mean `no header` but there is some problem with the LRA context.

You indeed get 412 back on both scenarios (no LRA header and LRA header which caller thinks is valid but is not). But caller did sent a header, so when gets 412 back he knows it was not valid.

My interpretation and caller only needs to react on status 412 (and not 410 and 412)

**Comment by xstefank** [comment]
TBH, I am also inclined to option 2. But what I am struggling with the distinction between the wrong content of the header (asdfasdfadsf) and expired header. Not sure whether this distinction can be significant for users. @rdebusscher WDYT?

**Comment by ochaloup** [comment]
Thanks @xstefank for the analysis. Based on it I'm inclined for option 1.
I think it's important to inform the user he did everything right but the system state changed meanwhile.

I think it's confusing for user to get `412 (Precondition Failed)` if he sends the correct header. Everything seems to be fine from his perspective but meanwhile somebody closes/cancels the LRA and he receives `412` which shouts to him "you did something wrong".

Another example I hit recently. There is defined really a short LRA timeout on `@LRA` annotation. The LRA spec defines that the starting LRA and registering the participant could be two separate steps. The HTTP call hits the LRA client library which handles the `@LRA` annotations. The LRA client library starts the LRA, there is some lag in processing and the registering of participant comes a while later. Now the `412` is returned as the LRA timed-out. The user's initial HTTP call gets the `412 Precondition Failed` and which indicates he did something wrong. If there is `410 Gone` returned then it guides much more (from my perspective) that the LRA was there but it's not now and there is probably some processing issue but not the issue with the initial call.

**Comment by xstefank** [comment]
ok, I think what you've said makes sense. @rdebusscher can you please respond also?

**Comment by rdebusscher** [comment]
>if he sends the correct header. Everything seems to be fine from his perspective but meanwhile somebody closes/cancels the LRA and he receives 412 which shouts to him "you did something wrong".

412 just means the LRA context isn't valid, not that the developer did something wrong.

And is the developer going to respond differently to the fact if he receives 410 or 412?  When he didn't send a header, he knows it is required. When he did send one and receives 412 he knows it is no longer valid (unless he deliberately used some random value)

**Comment by mmusgrov** [comment]
All we need to do is:

1. Check that the spec says `410 Gone` means the implementation either does not know about this LRA or knows it has already finished.

2. With regards to `412 Precondition Failed` the javadoc for the annotation should indicate when that code can be returned to the caller.

So if 1 is not already clear we just add a clarifying sentence to the spec. And we should already be covered for 2.

**Comment by ochaloup** [comment]
@rdebusscher my point here is of trying to take the perspective of the user (and it's something, I think, that should be emphasized on many other places of the spec). The user usually knows nothing about what he does. He usually reads only one blogpost and start using the technology. He does not read javadoc or even spec.
(Often I'm not supported in this perspective how a normal user works. I had many discussions e.g. with @mmusgrov on this aspect where the arguments stood at the point that user has to read the javadoc.)

My perspective is here still the same. The API should be the most informative as it's possible.

My point of view is here
`412` -> the input for the call is somehow, currently we may just check the wrong header
`410` ->  the input was right, the user processed the initial call fine, but the system informs the LRA is not there

This is how I understand the HTTP error code meaning and how the REST API should grasp them (where I could be misinterpreting something).

What is, please, your perspective on this point of view?

**Comment by rdebusscher** [comment]
>The user usually knows nothing about what he does. He usually reads only one blogpost and start using the technology. He does not read javadoc or even spec.

Will he know then the difference between 410 and 412?

This discussion is clearly a difference in taste. My point was to make it as simple as possible for the end user.

But I'm fine if everyone else chooses for status 410 and 412.

**Comment by ochaloup** [comment]
@rdebusscher that's a good point. My assumption is easily disregarded in pointingthat the `410` and `412` could be understood as a similar output by user.
Still, my approach (plus, I experienced a similar way of others) is to call something, get an error, check what the HTTP error means and then search about the meaning for the technology. Then the `Gone` against `Precondition Failed` give me a different "feeling" about the error.

In summary, I still incline for the option 1. But taking option 2 with informing in the documentation that the `412` means as well timeout is fine with me as well.

**Comment by xstefank** [comment]
So do we agree to go with option 1 for now? If this will be too burdensome for our users we will remove 410 from this use case in LRA 2.0. 

This is the last outstanding issue for LRA 1.0 so let's finish this one and move on ;)

**Comment by mmusgrov** [comment]
> So do we agree to go with option 1 for now? If this will be too burdensome for our users we will remove 410 from this use case in LRA 2.0.
> 
> This is the last outstanding issue for LRA 1.0 so let's finish this one and move on ;)

So what about my last (and first) comment? How invasive will your change be? Won't too much extra text obfuscate? Continual addition of "clarifying text" scattered throughout is making our spec difficult to parse and it is already difficult to discover what the semantics of the protocol/behaviour should be.

**Comment by xstefank** [comment]
@mmusgrov I will still be doing one final read/clean up after we'll finish all the issues. I don't think that this would be extra text. But we would need to add 410 to possible outcomes of LRA invocations somewhere, probably a new table. I don't think that 412 cases are covered elsewhere than in the LRA annotation javadoc.

But first we need an agreement on the solution for this issue.

**Comment by mmusgrov** [comment]
> @mmusgrov I will still be doing one final read/clean up after we'll finish all the issues. I don't think 

The text clean up could be substantial, we've all been chipping away at the text and we all have different grammatical styles so will no doubt want conflicting changesets.

> that this would be extra text. But we would need to add 410 to possible outcomes of LRA 
> invocations somewhere, probably a new table. I don't think that 412 cases are covered elsewhere than in the LRA annotation javadoc.

Updating the table sounds like a useful addition and I would vote yes to the issue on that basis.

> 
> But first we need an agreement on the solution for this issue.

Well it would depend on what text you add. My opinion is that we just need to verify that we say somewhere what `410 Gone` means.

**Comment by xstefank** [comment]
@mmusgrov we are saying what 410 means but only for participant return values I believe.

**Comment by mmusgrov** [comment]
> @mmusgrov we are saying what 410 means but only for participant return values I believe.

Well those are already well defined, for example the Compenstate return value is [here](https://github.com/eclipse/microprofile-lra/blob/master/api/src/main/java/org/eclipse/microprofile/lra/annotation/Compensate.java#L127).

It is probably easier if you just raise a PR which indicates where we aren't being precise enough.

**Comment by xstefank** [comment]
Agreed, PR will be clear. But I am not sure we've reached an agreement on the solution so far. If yes, then I will raise a PR to have both return values 412 and 410 (Option 1).

**Comment by mmusgrov** [comment]
> Agreed, PR will be clear. But I am not sure we've reached an agreement on the solution so far. If yes, then I will raise a PR to have both return values 412 and 410 (Option 1).

Can you either raise a PR or restate which problem you want to solve.

**Comment by rdebusscher** [comment]
I don't see the need for a differentiation in 410 and 412 here (no header and header with wrong or inactive LRAid). But since I'm the only one. go ahead, it is not a major point.

**Comment by xstefank** [comment]
@rdebusscher thanks for confirming Rudy.
